File: src/geometry/tools/lines/pointOnLine.ts
import { orient2dVector } from '../../index.js';

import type {
  MValue,
  Properties,
  VectorFeature,
  VectorLineString,
  VectorLineStringGeometry,
  VectorPoint,
} from '../../index.js';

/**
 * Check to see if a point is on a line. Uses predicates to ensure the point is truly on the line
 * @param line - the line to check against
 * @param point - the point to check if it is on the line
 * @param epsilon - the buffer to use to check if the point is on the line within epsilon. Defaults to 0
 * @returns True if the point is on the line
 */
export function pointOnLine<
  M = Record<string, unknown>,
  D extends MValue = Properties,
  P extends Properties = Properties,
>(
  line:
    | VectorLineString<D>
    | VectorLineStringGeometry<D>
    | VectorFeature<M, D, P, VectorLineStringGeometry<D>>,
  point: VectorPoint,
  epsilon: number = 0,
): boolean {
  const { min, max, abs } = Math;
  const vectorLines: VectorLineString<D> =
    'geometry' in line
      ? line.geometry.coordinates
      : 'coordinates' in line
        ? line.coordinates
        : line;

  if (vectorLines.length < 2) return false;

  for (let i = 0; i < vectorLines.length - 1; i++) {
    // check if in bounding box of each segment
    const a = vectorLines[i];
    const b = vectorLines[i + 1];
    if (
      point.x >= min(a.x, b.x) &&
      point.x <= max(a.x, b.x) &&
      point.y >= min(a.y, b.y) &&
      point.y <= max(a.y, b.y)
    ) {
      // lastly check if the point is on the segment
      const cross = orient2dVector(a, b, point);
      if (abs(cross) <= epsilon) return true;
    }
  }

  return false;
}

File: rust/geometry/tools/lines/point_on_line.rs
use crate::geometry::orient2d;
use libm::{fabs, fmax, fmin};
use s2json::GetXY;

/// Check to see if a point is on a line. Uses predicates to ensure the point is truly on the line
///
/// ## Parameters
/// - `line`: the line to check against
/// - `point`: the point to check if it is on the line
/// - `epsilon`: the buffer to use to check if the point is on the line within epsilon. Defaults to 0
///
/// ## Returns
/// True if the point is on the line
pub fn point_on_line<P: GetXY, Q: GetXY>(line: &[P], point: &Q, eps: Option<f64>) -> bool {
    let eps = eps.unwrap_or(0.0);

    if line.len() < 2 {
        return false;
    }

    let mut i = 0;
    while i < line.len() - 1 {
        // check if in bounding box of each segment
        let a = &line[i];
        let b = &line[i + 1];
        if point.x() >= fmin(a.x(), b.x())
            && point.x() <= fmax(a.x(), b.x())
            && point.y() >= fmin(a.y(), b.y())
            && point.y() <= fmax(a.y(), b.y())
        {
            // lastly check if the point is on the segment
            let cross = orient2d(a.x(), a.y(), b.x(), b.y(), point.x(), point.y());
            if fabs(cross) <= eps {
                return true;
            }
        }

        i += 1;
    }

    false
}
